src/Collections/Sequences/NatSeq.dfy

```diff
@@ -0,0 +1,602 @@
+// RUN: %dafny /compile:0 /noNLarith "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*******************************************************************************
+*  Original: Copyright (c) 2020 Secure Foundations Lab
+*  SPDX-License-Identifier: MIT
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+/* The first element of a sequence is the least significant word; the last
+element is the most significant word. */
+
+include "../../NonlinearArithmetic/DivMod.dfy"
+include "../../NonlinearArithmetic/Mul.dfy"
+include "../../NonlinearArithmetic/Power.dfy"
+include "Seq.dfy"
+
+abstract module NatSeq {
+
+  import opened DivMod
+  import opened Mul
+  import opened Power
+  import opened Seq
+
+  /* Upper bound of a word */
+  function method BOUND(): nat
+    ensures BOUND() > 1
+
+  type uint = i: int | 0 <= i < BOUND()
+
+  //////////////////////////////////////////////////////////////////////////////
+  //
+  // ToNat definition and lemmas
+  //
+  //////////////////////////////////////////////////////////////////////////////
+
+  /* Converts a sequence to a nat beginning from the least significant word. */
+  function method {:opaque} ToNat(xs: seq<uint>): nat
+  {
+    if |xs| == 0 then 0
+    else
+      LemmaMulNonnegativeAuto();
+      ToNat(DropFirst(xs)) * BOUND() + First(xs)
+  }
+
+  /* Converts a sequence to a nat beginning from the most significant word. */
+  function method {:opaque} ToNatRev(xs: seq<uint>): nat
```

Comment:
I wouldn't use the "Rev" suffix as I incorrectly assumed it meant this function would behave like `ToNat(Seq.Reverse(xs))`. Perhaps these two functions could be `ToNatLeft` and `ToNatRight`?

In fact, now I'm curious whether these could use `Seq.FoldLeft` and `Seq.FoldRight` instead. If they can't, that feels like an indication we need more lemmas on those Fold functions for them to be actually useful. :)